---
title: "Exploring sequence and phylogenetic diversity for a taxonomic group of interest."
author: 
- name: Nathan D. Olson
  affiliation: National Institute of Standards and Technology, Maryland, USA
package: metagenomeFeatures
output: 
    BiocStyle::html_document:
        toc_float: true
vignette: >
  %\VignetteIndexEntry{Exploring sequence and phylogenetic diversity for a taxonomic group of interest.}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
suppressPackageStartupMessages(library(metagenomeFeatures)) 
suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(ggtree))
```

# Vignette overview
16S rRNA reference databases are most often used to annotate OTUs or sequence variants in experimental dataset. 
The reference databases can also be used to gain study specific taxonomic groups. 
`MgDb` objects for 16S rRNA sequence reference databases can be used to explore the sequence and phylogenetic diversity of taxonomic groups of interest. 
In this vignette we will explore the 16S rRNA diversity for the _Gammaproteobaceria_ class using the Greengenes 16S rRNA database version 13.8 clustered at 85\% similarity. 
A `MgDb` object with the greengenes database version 13.8 85\% similarity OTUs (`gg85`) is included in the `r Biocpkg("metagenomeFeatures")` package.  

Other databases are avilable as [Bioconductor AnnotationData](https://www.bioconductor.org/packages/release/data/annotation/).  

- Greengenes Release 13.5: `r Biocpkg("greengenesMgDb13.5")`  
- Greengenes Release 13.8 99% OTUs: `r Biocpkg("greengenes13.8_99MgDb")`  
- Ribosomal Database Project Release 11.5: `r Biocpkg("ribosomaldatabaseproject11.5MgDb")`  
- Silva 128.1: `r Biocpkg("silva128.1MgDb")`

We will first load the `MgDb` class object, then select the taxa of interest, and finally explore the phylogenetic, and taxonomic composition of the _Enterobacteriace_ family. 
In addition to using the `r Biocpkg("metagenomeFeatures")` package, the vignette uses `r CRANpkg("dplyr")` and `r CRANpkg("tidyr")` for data manipulation, `r CRANpkg("ggplot2")` and `r Biocpkg("ggtree")` for data visualization.


```{r eval=FALSE}
library(metagenomeFeatures)
library(tidyverse)
library(ggtree)
```

# Loading and subsetting the database 
The `gg85` database is loaded using `get_gg13.8_85MgDb()`.  

```{r}
gg85 <- get_gg13.8_85MgDb()
```  

Next the `mgDb_select()` function is first used to subset the database, the `key` arguments is used to define the taxonmic group(s) of interest and `keytype` is used to define the taxonomic level for the group(s) of interest. 
With the subsetted database you can analyze the taxonomy, sequences, and phylogeny for the taxonomic group of interest. 
The select function returns single object with a subset of the database taxonomic, sequence, or phylogenetic data, as well as a named list with any two or all three data types.  

```{r}
gamma_16S <- mgDb_select(gg85, 
                          type = "all", 
                          keys = "Gammaproteobacteria", 
                          keytype = "Class")
```

# Taxonomic Analysis 
We want to know how many genera in the Class Gammaproteobacteria there are sequences for in the Greengenes 85% OTU database as well as how many sequences are assigned to each genera. 
We will first create a data frame with the desired information then plot the results. 

```{r}
## Per genus count data
gamma_df <- gamma_16S$taxa %>% 
    group_by(Genus) %>% 
    summarise(Count = n()) %>% 
    ungroup() %>% 
    mutate(Genus = fct_reorder(Genus, Count)) 

## Count info for text 
escherichia_count <- gamma_df$Count[gamma_df$Genus == "g__Escherichia"]

## excluding unassigned genera and genera with only one assigned sequence
gamma_trim_df <- gamma_df %>% filter(Genus != "g__", Count > 1)
```

For the Greengenes database there are a total of `r sum(gamma_df$Count)` OTUs assigned to `r nlevels(gamma_df$Genus)` genera in the Class Gammaproteobacteria. 
The number of OTUs assigned to specific genera, range from `r sort(gamma_df$Count, decreasing = TRUE)[2]` to `r min(gamma_df$Count)` (Fig. \@ref(fig:generaCount)). 
As this database is preclustered to 85\% similarity the number of OTUs per genus is more of an indicator of genera diversity than how well a genus is represented in the database. 
For example no sequences present in the database are assigned to the genus _Shigella_ and only `r escherichia_count` are assigned to _Escherichia_. 
OTUs only assigned to the family, `g__`, is the most abundant group, `r gamma_df$Count[gamma_df$Genus == "g__"]`.   

__TODO__: interesting genera for a narrative

```{r generaCount, fig.cap = "Number of seqeunces assigned to genera in the Class Gammaproteobacteria. Only Genera with more than one assigned sequence are shown.", fig.width = 6, fig.height = 6} 
 ggplot(gamma_trim_df) + 
    geom_bar(aes(x = Genus, y = Count), stat = "identity") + 
    labs(y = "Number of OTUs") + 
    coord_flip() + 
    theme_bw() 
```


# Phylogenetic Diversity  
We will use the `r Biocpkg("ggtree")` package to visualize the phylogenetic tree and annotate the tips with Genera information. 
Of the genera with more than 3 representative OTUs _Stenotrophomonas_ is the only genus with unclassified and other OTUs within the clade (Fig. \@ref(fig:annoTree)). 

```{r annoTree, fig.cap = "Phylogenetic tree of Gammaproteobacteria class OTU representative sequences.", fig.height = 8, fig.width = 6, message = FALSE}
library(ggtree)

genus_lab <- paste0("g__", c("","Stenotrophomonas", "Pseudomonas"))
genus_anno_df <- gamma_16S$taxa %>% 
    group_by(Genus) %>% 
    mutate(Count = n()) %>% 
    ungroup() %>% 
    mutate(Genus_lab = if_else(Count > 3, Genus, ""))
    # mutate(Genus_lab = if_else(Genus %in% genus_lab, Genus, ""),
    #        Genus_lab = factor(Genus_lab, levels = c("", genus_lab)))

ggtree(gamma_16S$tree) %<+% 
    genus_anno_df + 
    ## Add taxa name for unlabeled Stenotrophomonas branch
    geom_tippoint(aes(color = Genus_lab), size = 3) + 
    scale_color_manual(values = c("#FF000000","darkorange", "blue", "green", "tan", "black")) +
    theme(legend.position = "bottom") + 
    labs(color = "Genus")
```

# Sequence Diversity 
The `mgDb_select()` function returns the subsetted sequence data as a biostring object.
The sequence data can be used for additional analysis such as, multiple sequencing alignment, primer region extraction, or PCR primer design. 

```{r}
gamma_16S$seq
```

# Session info {.unnumbered}
```{r sessionInfo, echo=FALSE}
sessionInfo()
```
