---
title: "Example 16S Annotation Workflow"
author: "Nathan D. Olson"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

```{r, echo=FALSE, message=FALSE}
library(metagenomeFeatures)
library(greengenes13.5MgDb)
```

## Overview
A demonstration of how to use the metagenomeFeatures package for taxonimic annotation 16S marker gene sequence data.  
- Retrieve taxonomic information for a set of database Key values and generate a metagenomeAnnotation class object
- Split metagenomeAnnotation class object for a user defined taxonomic level
- Summary table and graphs for taxonomic information.


## Retrieve taxonomic information for a set of sequences.
```{r}
query_data <- data.frame(Keys = c("4324716", "246960", "222675", "156874", "4383832",
            "4383502", "315344", "2655590", "552241", "4327819",
            "328031", "4452386", "259209", "110470", "515223",
            "3453905", "4363471", "329924", "4373197", "3755712"),
                        otu_counts = 1:20)
MgAnnoDF <- annotate(gg13.5MgDb, query_df = query_data)
MgAnnoDF
```

## metagenomeAnnotation split_by method
```{r}
split_by(MgAnnoDF, "Phylum")
```

```{r}
sessionInfo()
```
