---
title: "Exploring a MgDb Object"
author: "Nathan D. Olson"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

The MgDb Class in the metagenomeFeatures package includes the sequences and taxonomic information for a 16S database.  The following vignette demonstrates the class methods for exploring and subsetting a MgDb object using the greengenes13.5MgDb annotation package. 

## greengenes13.5MgDb Package
The greengenes13.5MgDb package exports a MgDb class object, gg13.5MgDb, with sequence and taxonomic database information.

```{r}
library(greengenes13.5MgDb)
gg13.5MgDb
```


## MgDb Methods
taxa_keytypes
```{r}
gg13.5MgDb$taxa_keytypes()
```


```{r}
gg13.5MgDb$taxa_columns()
```

```{r}
head(gg13.5MgDb$taxa_keys(keytype = c("Kingdom")))
```


## Select Methods
Used to retrieve db entries for a specified taxanomic group or id list, can return either taxonomic, sequences information, or both. 

### Selecting taxonomic information
```{r}
gg13.5MgDb$select(type = "taxa",
                keys = c("Vibrio", "Salmonella"),
                keytype = "Genus")
```

### Selecting sequence information
```{r}
gg13.5MgDb$select(type = "seq",
                keys = c("Vibrio", "Salmonella"),
                keytype = "Genus")
```



### Selecting both
```{r}
gg13.5MgDb$select(type = "both",
                keys = c("Vibrio", "Salmonella"),
                keytype = "Genus")
```


```{r}
sessionInfo()
```

