---
title: "Using metagenomeFeatures to Retrieve Feature Data."
author: 
- name: Nathan D. Olson
  affiliation: National Institute of Standards and Technology, Maryland, USA
package: metagenomeFeatures
output: 
    BiocStyle::html_document:
        toc_float: true
vignette: >
  %\VignetteIndexEntry{Using metagenomeFeatures to Retrieve Feature Data.}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
references:
- id: rousk2010soil
  title: Soil bacterial and fungal communities across a pH gradient in an arable soil
  author:
  - family: Rousk
    given: Johannes
  - family: B{\aa}{\aa}th
    given: Erland
  - family: Brookes
    given: Philip C
  - family: Lauber
    given: Christian L
  - family: Lozupone
    given: Catherine
  - family: Caporaso
    given: J Gregory
  - family: Knight
    given: Rob
  - family: Fierer
    given: Noah
  container-title: The ISME journal
  volume: 4
  issue: 10
  publisher: Nature Publishing Group
  page: 1340
  type: article-journal
  issued:
    year: 2010
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  warning = FALSE
)
suppressPackageStartupMessages(library(metagenomeFeatures)) 
suppressPackageStartupMessages(library(phyloseq)) 
```

`r Biocpkg("metagenomeFeatures")` and associated annotation packages can be used to obtain phylogentic trees, and representative sequences for 16S rRNA marker gene sequences for experimental datasets when closed reference clustering is used. 
[QIITA](https://qiita.ucsd.edu) is a public repository for sharing OTU tables and raw sequence data from 16S rRNA marker gene studies, however some of the datasets do not include representative sequences or phylogenetic trees. 
In this vignette we will demostrate using `metagenomeFeatures` and the Greengenes 16S rRNA database version 13.8 85\% OTUs to obtain a phylogenetic tree and representative sequences for Rousk et al. [-@rousk2010soil], a soil microbiome study obtained from QIITA. 
The tree and sequence data will then be used to generate a `phyloseq-class` object for community analysis. 
The `r Biocpkg("phyloseq")` package defines a data structure, `phyloseq-calss`, and for working with 16S rRNA experimental data.

# Greengenes 13.8 85\% OTU MgDb
The gg 13.8 85% OTU is provided as part of the `r Biocpkg("metagenomeFeatures")` package. 
`gg85` is a `MgDb-class` object with the taxonomic heirarchy, sequence data, and phylogeny for the Greengenes database clustered at the 0.85 similarity threshold. 

```{r message = FALSE}
library(metagenomeFeatures)
gg85 <- get_gg13.8_85MgDb()
```

See the "metagenomeFeatures classes and methods" vignette, `Vignettes("metagenomeFeatures classes and methods")` for more information on `gg85` and methods for working with `MgDb-class` objects. 

##  QIITA Dataset 
For this vignette we are using 16S rRNA data from a soil microbiome study, https://qiita.ucsd.edu/study/description/94 [@rousk2010soil]. 
A BIOM and qiime mapping file for the study can be obtained from QIITA. 
A vector of Greengenes for the study cluster centers is included in this package for use in this vignette. 

```{r}
data(qiita_study_94_gg_ids)
soil_gg_ids <-  qiita_study_94_gg_ids
```

## Obtaining Sequences and Phylogenetic Tree 
```{r}
soil_mgF <- annotateFeatures(gg85, soil_gg_ids)
```

The resulting `mgFeatures` class object has the taxonomic heirarchy, phylogeny, and sequence data for the study OTUs. 
```{r}
soil_mgF

# Sequence data
mgF_seq(soil_mgF)

# Tree data 
mgF_tree(soil_mgF)
```

# Using `mgFeatures-class` object with `phyloseq` 
We will use the phyogenetic tree from the `mgFeatures-object` along with the OTU table for a beta diversity analysis.
```{r}
library(phyloseq)
```

Load abundance an taxonomy data from QITTA biom file
```{r}
# biom_file <-  file.path(data_dir, "229_otu_table.biom")
biom_file <- "inst/extdata/229_otu_table.biom"
soil_ps <- phyloseq::import_biom(BIOMfilename = biom_file)

## Adding sample data
## TODO - move to inst/scripts
# library(tidyverse)
# sample_dat <- read_tsv("~/Downloads/mapping_files/2301_mapping_file.txt") %>%
#     dplyr::rename(SampleID = `#SampleID`) %>%
#     select(SampleID, carb_nitro_ratio, ph, tot_nitro, tot_org_carb)
# write_tsv(sample_dat, "inst/extdata/229_sample_data.tsv")
sample_dat <- read.delim("inst/extdata/229_sample_data.tsv")
rownames(sample_dat) <- sample_dat$SampleID
sample_data(soil_ps) <- sample_dat
```

```{r}
soil_ps
```

the `gg85` on has a subset of the representative sequences in the soil_mgF, 
subsetting soil_ps to only include OTUs in `gg85`.
```{r}
soil_tree <- mgF_tree(soil_mgF)
soil_ps_gg85 <- prune_taxa(taxa = soil_tree$tip.label, x = soil_ps)
# Removing samples with no OTUs in `gg85`
soil_ps_gg85 <- prune_samples(samples = sample_sums(soil_ps_gg85) != 0,
                              x = soil_ps_gg85)
```


Use `soil_mgF` to define the `seq` and `tree` slots
```{r}
## Defining tree slot
phy_tree(physeq = soil_ps_gg85) <- soil_tree

## Defining seq slot
soil_ps_gg85@refseq <- mgF_seq(soil_mgF)
```


The phyloseq class can then be used for microbial community analysis requiring a phylogenetic tree such as phylogenetic beta diversity. 
For the subset of features used here pH is a primary driver for differences in beta diversity between the samples. 
Nearly 30\% of the total variation in the data is explained by the first axis and sample pH is correlated with the first axis. 

```{r}
soil_ord <- ordinate(physeq = soil_ps_gg85, distance = "wunifrac", method = "PCoA")
plot_ordination(soil_ps_gg85, soil_ord, color = "ph", 
                type="sample", 
                title="pH - Weighted Unifrac PCoA")
```

# Session info {.unnumbered}
```{r sessionInfo, echo=FALSE}
sessionInfo()
```



